control-plane/csi-controller/src/controller.rs

```diff
@@ -0,0 +1,726 @@
+
+        // For exaplanation of accessibilityRequirements refer to a table at
+        // https://github.com/kubernetes-csi/external-provisioner.
+        // Our case is WaitForFirstConsumer = true, strict-topology = false.
+        //
+        // The first node in preferred array the node that was chosen for running
+        // the app by the k8s scheduler. The rest of the entries are in random
+        // order and perhaps don't even run mayastor csi node plugin.
+        //
+        // The requisite array contains all nodes in the cluster irrespective
+        // of what node was chosen for running the app.
+        let mut allowed_nodes: Vec<String> = Vec::new();
+        let mut preferred_nodes: Vec<String> = Vec::new();
+
+        if let Some(reqs) = args.accessibility_requirements {
+            for r in reqs.requisite.iter() {
+                for (k, v) in r.segments.iter() {
+                    // We are not able to evaluate any other topology requirements than
+                    // the hostname req. Reject all others.
+                    if k != K8S_HOSTNAME {
+                        return Err(Status::invalid_argument(
+                            "Volume topology other than hostname not supported",
+                        ));
+                    }
+                    allowed_nodes.push(v.to_string());
+                }
+            }
+
+            for p in reqs.preferred.iter() {
+                for (k, v) in p.segments.iter() {
+                    // Ignore others than hostname (it's only preferred)
+                    if k == K8S_HOSTNAME {
+                        preferred_nodes.push(v.to_string());
+                    }
+                }
+            }
+        }
+
+        let u = Uuid::parse_str(&volume_uuid).map_err(|_e| {
+            Status::invalid_argument(format!("Malformed volume UUID: {}", volume_uuid))
+        })?;
+
+        // First check if the volume already exists.
+        if let Some(existing_volume) = MayastorApiClient::get_client()
+            .list_volumes()
+            .await?
+            .into_iter()
+            .find(|v| v.spec.uuid == u)
+        {
+            check_existing_volume(&existing_volume, replica_count, size, pinned_volume)?;
+            debug!(
+                "Volume {} already exists and is compatible with requested config",
+                volume_uuid
+            );
+        } else {
+            MayastorApiClient::get_client()
+                .create_volume(
+                    &volume_uuid,
+                    replica_count,
+                    size,
+                    &allowed_nodes,
+                    &preferred_nodes,
+                )
+                .await?;
+
+            debug!(
+                "Volume {} successfully created, pinned volume = {}",
+                volume_uuid, pinned_volume
+            );
+        }
+
+        let vt_mapper = VolumeTopologyMapper::init().await?;
```

Comment:
Should this be called before creating the volume? Otherwise if it fails we may "leak" the volume?

Reply:
Yup, good point.